In [1]:
import pandas as pd
pd.options.display.max_columns = 99

## Selection of chunk size
In this step we will assume that we have only memory of 10 megabytes. So we will be using half of this place for data loading and other half will leave empty for operations.
So we will start by first loading the data size of nrows=1000 then by increasing or decreasing we will check the optimum value for chunksize.
So lets get started

In [2]:
loans = pd.read_csv('loans_2007.csv',nrows=1000)
loans.memory_usage(deep=True).sum()/1024**2

1.5506629943847656

In [3]:
loans = pd.read_csv('loans_2007.csv',nrows=5000)
loans.memory_usage(deep=True).sum()/1024**2

7.7490835189819336

In [4]:
loans = pd.read_csv('loans_2007.csv',nrows=2000)
loans.memory_usage(deep=True).sum()/1024**2

3.1007261276245117

In [5]:
loans = pd.read_csv('loans_2007.csv',nrows=3000)
loans.memory_usage(deep=True).sum()/1024**2

4.6512298583984375

Now we have found the optimum size for number of sample in each chunk as 3000. So lets focus on the datatype of columns and their 1st value so we will be able to identify if there are some columns available or not for optimization.

In [6]:
loans.dtypes

id                              int64
member_id                     float64
loan_amnt                     float64
funded_amnt                   float64
funded_amnt_inv               float64
term                           object
int_rate                       object
installment                   float64
grade                          object
sub_grade                      object
emp_title                      object
emp_length                     object
home_ownership                 object
annual_inc                    float64
verification_status            object
issue_d                        object
loan_status                    object
pymnt_plan                     object
purpose                        object
title                          object
zip_code                       object
addr_state                     object
dti                           float64
delinq_2yrs                   float64
earliest_cr_line               object
inq_last_6mths                float64
open_acc    

In [7]:
# number of string type columns
loans.select_dtypes(include=['object']).columns.shape[0]

21

In [8]:
# number of numeric type columns
loans.select_dtypes(exclude=['object']).columns.shape[0]

31

### Here we have found that this dataframe contains 31 numeric columns while 21 object columns
**NOTE: We can use boolean values as 1 for True and 0 for False.**

In [9]:
loans.head()

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0  1077501  1296599.0     5000.0       5000.0           4975.0   36 months   
1  1077430  1314167.0     2500.0       2500.0           2500.0   60 months   
2  1077175  1313524.0     2400.0       2400.0           2400.0   36 months   
3  1076863  1277178.0    10000.0      10000.0          10000.0   36 months   
4  1075358  1311748.0     3000.0       3000.0           3000.0   60 months   

  int_rate  installment grade sub_grade                 emp_title emp_length  \
0   10.65%       162.87     B        B2                       NaN  10+ years   
1   15.27%        59.83     C        C4                     Ryder   < 1 year   
2   15.96%        84.33     C        C5                       NaN  10+ years   
3   13.49%       339.31     C        C1       AIR RESOURCES BOARD  10+ years   
4   12.69%        67.79     B        B5  University Medical Group     1 year   

  home_ownership  annual_inc verification_status   issue_d  loan_status  \
0           RENT     24000.0            Verified  Dec-2011   Fully Paid   
1           RENT     30000.0     Source Verified  Dec-2011  Charged Off   
2           RENT     12252.0        Not Verified  Dec-2011   Fully Paid   
3           RENT     49200.0     Source Verified  Dec-2011   Fully Paid   
4           RENT     80000.0     Source Verified  Dec-2011      Current   

  pymnt_plan         purpose                 title zip_code addr_state    dti  \
0          n     credit_card              Computer    860xx         AZ  27.65   
1          n             car                  bike    309xx         GA   1.00   
2          n  small_business  real estate business    606xx         IL   8.72   
3          n           other              personel    917xx         CA  20.00   
4          n           other              Personal    972xx         OR  17.94   

   delinq_2yrs earliest_cr_line  inq_last_6mths  open_acc  pub_rec  revol_bal  \
0          0.0         Jan-1985             1.0       3.0      0.0    13648.0   
1          0.0         Apr-1999             5.0       3.0      0.0     1687.0   
2          0.0         Nov-2001             2.0       2.0      0.0     2956.0   
3          0.0         Feb-1996             1.0      10.0      0.0     5598.0   
4          0.0         Jan-1996             0.0      15.0      0.0    27783.0   

  revol_util  total_acc initial_list_status  out_prncp  out_prncp_inv  \
0      83.7%        9.0                   f       0.00           0.00   
1       9.4%        4.0                   f       0.00           0.00   
2      98.5%       10.0                   f       0.00           0.00   
3        21%       37.0                   f       0.00           0.00   
4      53.9%       38.0                   f     461.73         461.73   

    total_pymnt  total_pymnt_inv  total_rec_prncp  total_rec_int  \
0   5863.155187          5833.84          5000.00         863.16   
1   1008.710000          1008.71           456.46         435.17   
2   3005.666844          3005.67          2400.00         605.67   
3  12231.890000         12231.89         10000.00        2214.92   
4   3581.120000          3581.12          2538.27        1042.85   

   total_rec_late_fee  recoveries  collection_recovery_fee last_pymnt_d  \
0                0.00        0.00                     0.00     Jan-2015   
1                0.00      117.08                     1.11     Apr-2013   
2                0.00        0.00                     0.00     Jun-2014   
3               16.97        0.00                     0.00     Jan-2015   
4                0.00        0.00                     0.00     Jun-2016   

   last_pymnt_amnt last_credit_pull_d  collections_12_mths_ex_med  \
0           171.62           Jun-2016                         0.0   
1           119.66           Sep-2013                         0.0   
2           649.91           Jun-2016                         0.0   
3           357.48           Apr-2016                     

### Selection of data types
Here we can notice that id and member_id dosen't have any strong correlatin with any point or it can't show causatin on the basis of just id and member_id so we can drop these two values.
***
Again after checking other rows we can see int_rate and revol_util also has string values but can be converted to numeric data types.
***
Also issue_d and earliest_cr_line can be converted into datetime object while parsing those columns.
***
If we give more thought on this data we can see we can eaily drop title column because it is just extra information for purpose column also we can drop zip code because we have already a variable in address which shows same information in string value.

In [10]:
# the columns which we would like not to select
drop_cols = ['id','member_id','title','zip_code']
# the columns which we would like to convert to numeric
num_con_cols = ['int_rate','revol_util']
# the columns which we would like to convert to datetime64
datetime_con_cols = ['issue_d','earliest_cr_line']

Before doing any processes let's analyze all data first, so we can check if we are on right path or not.

In [11]:
# numeric columns
num_cols = loans.select_dtypes(include=['float','int']).columns
# object columns
obj_cols = loans.select_dtypes(include=['object']).columns
for col in num_cols:
    print(col)
print('----------------------------------------')
for col in obj_cols:
    print(col)

id
member_id
loan_amnt
funded_amnt
funded_amnt_inv
installment
annual_inc
dti
delinq_2yrs
inq_last_6mths
open_acc
pub_rec
revol_bal
total_acc
out_prncp
out_prncp_inv
total_pymnt
total_pymnt_inv
total_rec_prncp
total_rec_int
total_rec_late_fee
recoveries
collection_recovery_fee
last_pymnt_amnt
collections_12_mths_ex_med
policy_code
acc_now_delinq
chargeoff_within_12_mths
delinq_amnt
pub_rec_bankruptcies
tax_liens
----------------------------------------
term
int_rate
grade
sub_grade
emp_title
emp_length
home_ownership
verification_status
issue_d
loan_status
pymnt_plan
purpose
title
zip_code
addr_state
earliest_cr_line
revol_util
initial_list_status
last_pymnt_d
last_credit_pull_d
application_type


### Checking the length of dataframe

In [12]:
# creating iteration object for each chunk of object cols
chunk_iter = pd.read_csv('loans_2007.csv',chunksize=3000,\
                        usecols=obj_cols)

In [13]:
df_length = 0
for chunk in chunk_iter:
    df_length += chunk.shape[0]
df_length

42538

** The length of whole dataframe is 42538.**

Now lets' check how many unique values each object column has.

In [14]:
chunk_iter = pd.read_csv('loans_2007.csv',chunksize=3000,\
                        usecols=obj_cols)
vc_dict = {}
for chunk in chunk_iter:
    for obj in obj_cols:
        vc_chunk = chunk[obj].value_counts()
        if obj in vc_dict:
            vc_dict[obj].append(vc_chunk)
        else:
            vc_dict[obj] = [vc_chunk]

In [56]:
obj_unique_dict = {}
for key,value in vc_dict.items():
    items = pd.concat(value)
    unique_count = items.groupby(items.index).sum().shape[0]
    obj_unique_dict[key] = unique_count
obj_unique_count = pd.Series(obj_unique_dict)
obj_unique_perc = obj_unique_count/df_length*100
obj_unique_df = pd.DataFrame([obj_unique_count,obj_unique_perc])
obj_unique_df = obj_unique_df.sort_values(axis=1,by=1,ascending=False)
obj_unique_df

emp_title         title   revol_util    zip_code  earliest_cr_line  \
0  30659.000000  21265.000000  1119.000000  837.000000        530.000000   
1     72.074381     49.990597     2.630589    1.967652          1.245945   

     int_rate  last_credit_pull_d  last_pymnt_d    issue_d  addr_state  \
0  394.000000          108.000000    103.000000  55.000000   50.000000   
1    0.926231            0.253891      0.242136   0.129296    0.117542   

   sub_grade    purpose  emp_length  loan_status     grade  home_ownership  \
0  35.000000  14.000000    12.00000     9.000000  7.000000        5.000000   
1   0.082279   0.032912     0.02821     0.021158  0.016456        0.011754   

   verification_status  pymnt_plan      term  initial_list_status  \
0             3.000000    2.000000  2.000000             1.000000   
1             0.007053    0.004702  0.004702             0.002351   

   application_type  
0          1.000000  
1          0.002351

In [ ]:
obj_

Here we can see that emp_title and title has most unique values which are nearly greater than 50%. So they are no use for our model.

In [16]:
chunk_iter = pd.read_csv('loans_2007.csv',chunksize=3000,\
                        usecols = num_cols)
null_num_dict = {}
for chunk in chunk_iter:
    for num in num_cols:
        null_count = chunk[num].isnull().sum()
        #print(null_count)
        if num in null_num_dict:
            null_num_dict[num] += null_count
        else:
            null_num_dict[num] = null_count
#null_num_dict

#### NULL VALUE COUNTS FOR NUMERIC COLUMNS

In [17]:
null_numerical_series = pd.Series(null_num_dict)
null_numerical_series.sort_values()

id                               0
member_id                        3
total_rec_int                    3
total_pymnt_inv                  3
total_pymnt                      3
revol_bal                        3
recoveries                       3
policy_code                      3
out_prncp_inv                    3
out_prncp                        3
total_rec_late_fee               3
loan_amnt                        3
last_pymnt_amnt                  3
total_rec_prncp                  3
funded_amnt_inv                  3
funded_amnt                      3
dti                              3
collection_recovery_fee          3
installment                      3
annual_inc                       7
inq_last_6mths                  32
total_acc                       32
delinq_2yrs                     32
pub_rec                         32
delinq_amnt                     32
open_acc                        32
acc_now_delinq                  32
tax_liens                      108
collections_12_mths_

#### Now we will calculate the memory usage for all chunks

In [18]:
chunk_iter = pd.read_csv('loans_2007.csv',chunksize=3000)
memory = 0
for chunk in chunk_iter:
    memory += chunk.memory_usage(deep=True).sum()/1024**2
memory

66.245146751403809

** 66.25 megabytes memory could be used if we load whole dataset in memory **

## Selection of columns haveing less than 50% unique values
As we can note that after sorting newly created **`obj_unique_df`** the columns which have less than 50% can be selected by just indexing.

In [58]:
convert_category = obj_unique_df.columns[2:].tolist()
convert_category

['revol_util',
 'zip_code',
 'earliest_cr_line',
 'int_rate',
 'last_credit_pull_d',
 'last_pymnt_d',
 'issue_d',
 'addr_state',
 'sub_grade',
 'purpose',
 'emp_length',
 'loan_status',
 'grade',
 'home_ownership',
 'verification_status',
 'pymnt_plan',
 'term',
 'initial_list_status',
 'application_type']

Also we have to do some cleaning on `int_rate and revol_util` columns as they are integers only but due to **%** symbole after them they are converted to string type. So let's define a function to do this task.

In [46]:
# function for converting string to integer
def convert_perc_int(string):
    string = string.strip()
    string = string.rstrip('%')
    return float(string)

In [45]:
s = '6.0%'
v = s.rstrip('%')
v=='6.0'
float('6.0')

6.0

Now that we have created function also the columns which we want to convert into category type. Lets' go on and see how much memory do we improve by these changes.<br>
Also we can parse dates in datetime object. We can find datetime type columns by analyzing the columns value

In [60]:
# first making a dictinoary for dtypes
dtypes = {}
for convert in convert_category:
    dtypes[convert] = 'category'

# selecting datetime type columns from object columns    
datetime_cols = ["issue_d", "earliest_cr_line", "last_pymnt_d", \
                 "last_credit_pull_d"]

In [68]:
chunk_iter = pd.read_csv('loans_2007.csv',chunksize=3000,dtype=dtypes,\
                         parse_dates=datetime_cols)
# memroy variable to calculate the total usage of memroy
memory = 0
# we have defined num_con_cols in very start of this notebook
for chunk in chunk_iter:
    for col in num_con_cols:
        chunk[col] = pd.to_numeric(chunk[col].apply(convert_perc_int))
    #chunk['id'] = chunk['id'].astype('int64')
    memory += chunk.memory_usage(deep=True).sum()/1024**2
memory

19.317028999328613

## We reduced total usage of memory from 66.25 megabytes to 19.32 megabytes which is more than 50% improvment

In [49]:
null_numerical_series.sort_values()

id                               0
member_id                        3
total_rec_int                    3
total_pymnt_inv                  3
total_pymnt                      3
revol_bal                        3
recoveries                       3
policy_code                      3
out_prncp_inv                    3
out_prncp                        3
total_rec_late_fee               3
loan_amnt                        3
last_pymnt_amnt                  3
total_rec_prncp                  3
funded_amnt_inv                  3
funded_amnt                      3
dti                              3
collection_recovery_fee          3
installment                      3
annual_inc                       7
inq_last_6mths                  32
total_acc                       32
delinq_2yrs                     32
pub_rec                         32
delinq_amnt                     32
open_acc                        32
acc_now_delinq                  32
tax_liens                      108
collections_12_mths_

In above series we can note that only id column has 0 null values which means we can only convert id column to integer column without any error. But it seems not a good idea for just one column

In [63]:
chunk.dtypes

id                                    object
member_id                            float64
loan_amnt                            float64
funded_amnt                          float64
funded_amnt_inv                      float64
term                                category
int_rate                             float64
installment                          float64
grade                               category
sub_grade                           category
emp_title                             object
emp_length                          category
home_ownership                      category
annual_inc                           float64
verification_status                 category
issue_d                       datetime64[ns]
loan_status                         category
pymnt_plan                          category
purpose                             category
title                                 object
zip_code                            category
addr_state                          category
dti       